File: support/handmade-cs/jsonReader.cs
// 以下是驗證程式碼, 不可使用
// using區段可能與實際使用的不一致, 要注意

using Newtonsoft.Json;

namespace Sheeter
{
    using Data_ = Handmade;
    using Key_ = System.String;
    using Store_ = Dictionary<string, Handmade>;

    /// <summary>
    /// $結構說明
    /// </summary>
    public partial class Handmade
    {
        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Pkey")]
        public int Pkey { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Skey")]
        public string Skey { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data1")]
        public bool Data1 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data2")]
        public bool[] Data2 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data3")]
        public int Data3 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data4")]
        public int[] Data4 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data5")]
        public long Data5 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data6")]
        public long[] Data6 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data7")]
        public float Data7 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data8")]
        public float[] Data8 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data9")]
        public double Data9 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data10")]
        public double[] Data10 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data11")]
        public string Data11 { get; set; }

        /// <summary>
        /// $欄位說明
        /// </summary>
        [JsonProperty("Data12")]
        public string[] Data12 { get; set; }
    }

    /// <summary>
    /// $結構說明
    /// </summary>
    public partial class HandmadeReader : Reader
    {
        /// <summary>
        /// 取得檔名物件
        /// </summary>
        public FileName FileName()
        {
            return new FileName("handmade", ".json");
        }

        /// <summary>
        /// 讀取資料
        /// </summary>
        public string FromData(string data)
        {
            try
            {
                Data = JsonConvert.DeserializeObject<Store_>(data);
            } // try
            catch
            {
                return "from data failed: deserialize failed";
            } // catch

            if (Data == null)
                return "from data failed: deserialize failed";

            return string.Empty;
        }

        /// <summary>
        /// 合併資料
        /// </summary>
        public string MergeData(string data)
        {
            Store_ tmpl;

            try
            {
                tmpl = JsonConvert.DeserializeObject<Store_>(data);
            } // try
            catch
            {
                return "merge data failed: deserialize failed";
            } // catch

            if (tmpl == null)
                return "merge data failed: deserialize failed";

            foreach (var itor in tmpl)
            {
                if (Data.ContainsKey(itor.Key))
                    return "merge data failed: key duplicate";

                Data[itor.Key] = itor.Value;
            } // for

            return string.Empty;
        }

        /// <summary>
        /// 清除資料
        /// </summary>
        public void Clear()
        {
            Data.Clear();
        }

        /// <summary>
        /// 取得資料
        /// </summary>
        public bool TryGetValue(Key_ key, out Data_ value)
        {
            return Data.TryGetValue(key, out value);
        }

        /// <summary>
        /// 檢查索引是否存在
        /// </summary>
        public bool ContainsKey(Key_ key)
        {
            return Data.ContainsKey(key);
        }

        /// <summary>
        /// 取得迭代器
        /// </summary>
        public IEnumerator<KeyValuePair<Key_, Data_>> GetEnumerator()
        {
            return Data.GetEnumerator();
        }

        /// <summary>
        /// 取得資料
        /// </summary>
        public Data_ this[Key_ key]
        {
            get { return Data[key]; }
        }

        /// <summary>
        /// 取得索引列表
        /// </summary>
        public ICollection<Key_> Keys
        {
            get { return Data.Keys; }
        }

        /// <summary>
        /// 取得資料列表
        /// </summary>
        public ICollection<Data_> Values
        {
            get { return Data.Values; }
        }

        /// <summary>
        /// 取得資料數量
        /// </summary>
        public int Count
        {
            get { return Data.Count; }
        }

        private Store_ Data = new Store_();
    }
}

File: support/handmade-go/jsonReader.go
// 以下是驗證程式碼, 不可使用

package sheeter

import (
	"encoding/json"
	"fmt"
)

// Handmade $結構說明
type Handmade struct {
	Pkey   int32     `json:"pkey"`   // $欄位說明
	Skey   string    `json:"skey"`   // $欄位說明
	Data1  bool      `json:"data1"`  // $欄位說明
	Data2  []bool    `json:"data2"`  // $欄位說明
	Data3  int32     `json:"data3"`  // $欄位說明
	Data4  []int32   `json:"data4"`  // $欄位說明
	Data5  int64     `json:"data5"`  // $欄位說明
	Data6  []int64   `json:"data6"`  // $欄位說明
	Data7  float32   `json:"data7"`  // $欄位說明
	Data8  []float32 `json:"data8"`  // $欄位說明
	Data9  float64   `json:"data9"`  // $欄位說明
	Data10 []float64 `json:"data10"` // $欄位說明
	Data11 string    `json:"data11"` // $欄位說明
	Data12 []string  `json:"data12"` // $欄位說明
}

// HandmadeReader $結構說明
type HandmadeReader struct {
	Data map[string]*Handmade // $欄位說明
}

// FileName 取得檔名物件
func (this *HandmadeReader) FileName() FileName {
	return NewFileName("handmade", ".json")
}

// FromData 讀取資料
func (this *HandmadeReader) FromData(data []byte) error {
	this.Data = map[string]*Handmade{}

	if err := json.Unmarshal(data, &this.Data); err != nil {
		return fmt.Errorf("from data: %w", err)
	} // if

	return nil
}

// MergeData 合併資料
func (this *HandmadeReader) MergeData(data []byte) error {
	tmpl := map[string]*Handmade{}

	if err := json.Unmarshal(data, &tmpl); err != nil {
		return fmt.Errorf("merge data: %w", err)
	} // if

	if this.Data == nil {
		this.Data = map[string]*Handmade{}
	} // if

	for k, v := range tmpl {
		if _, ok := this.Data[k]; ok {
			return fmt.Errorf("merge data: key duplicate")
		} // if

		this.Data[k] = v
	} // for

	return nil
}

// Clear 清除資料
func (this *HandmadeReader) Clear() {
	this.Data = nil
}

// Get 取得資料
func (this *HandmadeReader) Get(key string) *Handmade {
	if result, ok := this.Data[key]; ok {
		return result
	} // if

	return nil
}

// Keys 取得索引列表
func (this *HandmadeReader) Keys() (result []string) {
	for itor := range this.Data {
		result = append(result, itor)
	} // for

	return result
}

// Values 取得資料列表
func (this *HandmadeReader) Values() (result []*Handmade) {
	for _, itor := range this.Data {
		result = append(result, itor)
	} // for

	return result
}

// Count 取得資料數量
func (this *HandmadeReader) Count() int {
	return len(this.Data)
}
